src/nxt_cgroup.c

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright (C) Andrew Clayton
+ * Copyright (C) F5, Inc.
+ */
+
+#include <nxt_main.h>
+#include <nxt_process.h>
+#include <nxt_malloc.h>
+#include <nxt_fs.h>
+#include <nxt_cgroup.h>
+
+#include <stdio.h>
+#include <string.h>
+#include <limits.h>
+
+
+static char *nxt_mk_cgpath(const char *dir);
+
+
+nxt_int_t
+nxt_cgroup_proc_add(const nxt_task_t *task, nxt_process_t *process)
+{
+    int   ret, len;
+    FILE  *fp;
+    char  *cgpath;
+    char  cgprocs[PATH_MAX - 13];   /* - 13 for '/cgroup.procs' */
+
+    if (task->thread->runtime->type != NXT_PROCESS_MAIN
+        || nxt_process_type(process) != NXT_PROCESS_PROTOTYPE
+        || process->isolation.cgroup.path == NULL)
+    {
+        return NXT_OK;
+    }
+
+    cgpath = nxt_mk_cgpath(process->isolation.cgroup.path);
+    if (cgpath == NULL) {
+        return NXT_ERROR;
+    }
+    nxt_fs_mkdir_all((const u_char *) cgpath, 0777);
+
+    snprintf(cgprocs, sizeof(cgprocs), "%s/cgroup.procs", cgpath);
+    fp = fopen(cgprocs, "we");
+    if (fp == NULL) {
+        ret = NXT_ERROR;
+        goto out_free;
+    }
+
+    ret = NXT_OK;
+
+    setvbuf(fp, NULL, _IONBF, 0);
+    len = fprintf(fp, "%d\n", process->pid);
+    if (len < 0) {
+        ret = NXT_ERROR;
+    }
+
+    fclose(fp);
+
+out_free:
+
+    nxt_free(cgpath);
+
+    return ret;
+}
+
+
+void
+nxt_cgroup_cleanup(const nxt_process_t *process)
+{
+    char  *cgroot, *cgpath, *ptr;
+
+    cgroot = nxt_mk_cgpath("");
+    cgpath = nxt_mk_cgpath(process->isolation.cgroup.path);
+    if (cgroot == NULL || cgpath == NULL) {
+        goto out_free;
+    }
+
+    while (*cgpath != '\0' && strcmp(cgroot, cgpath) != 0) {
+        rmdir(cgpath);
+        ptr = strrchr(cgpath, '/');
+        *ptr = '\0';
+    }
+
+out_free:
+
+    nxt_free(cgroot);
+    nxt_free(cgpath);
+}
+
+
+static char *
+nxt_mk_cgpath(const char *dir)
+{
+    char    *path, *buf, *ptr;
+    size_t  pathlen;
+
+    path = buf = ptr = NULL;
+
+    /*
+     * If the path from the config is relative, we need to make
+     * the cgroup path include the main unit processes cgroup. I.e
+     *
+     *   NXT_CGROUP_ROOT/<main process cgroup>/<cgroup path>
+     */
+    if (dir[0] != '/') {
+        int   i;
+        FILE  *fp;
+
+        fp = fopen("/proc/self/cgroup", "re");
+        if (fp == NULL) {
+            return NULL;
+        }
+
+        fscanf(fp, "%ms", &buf);
+        fclose(fp);
+
+        if (buf == NULL) {
+            return NULL;
+        }
+
+        ptr = buf;
+        for (i = 0; i < 2; i++) {
+            ptr = strchr(ptr, ':');
+            if (ptr == NULL) {
+                goto out_free_buf;
+            }
+
+            ptr++;
+        }
+    }
+
+    pathlen = sizeof(char) * PATH_MAX;
```

Comment:
`sizeof(char) == 1` by definition.  What's this trying to tell?

Reply:
`pathlen` seems to be always `PATH_MAX`.  I'd just use that directly, no?

Reply:
I was flip-flopping whether to keep the sizeof() (I guess because you're normally doing 'sizeof(type) * n' for this kind of thing). It can go, no biggie...

Reply:
I guessed something like that.  For char I either use `sizeof(*buf) * n` or `n` directly, but `sizeof(char)` seems like blasfemy :D.  I think `n` directly will be better here.

Reply:
> pathlen seems to be always PATH_MAX. I'd just use that directly, no?

I was just thinking if we ever wanted pathlen to be something else... probably not and there's only two places in the code to change...

> I guessed something like that. For char I either use sizeof(*buf) * n or n directly, but sizeof(char) seems like blasfemy :D. I think n directly will be better here.

Sure and if I did it tomorrow I might just have done it that way!